pkg/chezmoi/archive.go

```diff
@@ -132,10 +134,44 @@ func isTarArchive(r io.Reader) bool {
 	return err == nil
 }
 
+func ImplicitDirHeader(dir string, fileHeader *tar.Header) *tar.Header {
+	return &tar.Header{
+		Typeflag: tar.TypeDir,
+		Name:     dir,
+		Mode:     0o755,
+		Size:     0,
+		Uid:      fileHeader.Uid,
+		Gid:      fileHeader.Gid,
+		Uname:    fileHeader.Uname,
+		Gname:    fileHeader.Gname,
+		ModTime:  fileHeader.ModTime,
+	}
+}
+
 // walkArchiveTar walks over all the entries in a tar archive.
 func walkArchiveTar(r io.Reader, f WalkArchiveFunc) error {
 	tarReader := tar.NewReader(r)
 	var skippedDirPrefixes []string
+	var seenDirs []string
```

Comment:
`seenDirs` should be a `map[string]struct{}` for fast lookups. Using `slice.Contains` does a linear search which makes the algorithm accidentally quadratic.

Reply:
I wasn't sure what the `struct{}` would be in this case? I opted for `map[string]bool` to keep it simple since I'm only using it as a flag.

Reply:
That would work, and I’m not sure whether `map[string]struct{}` would be better or exactly what it would be myself. At a guess, something other than `bool` would allow you to potentially put a `tar.Header` (the `implicitDirHeader` output?) or something similar in place.

Reply:
I think that @twpayne had a clearer idea on this when he wrote his original comment.

Reply:
`map[string]struct{}` is a weird Go idiomatic way of representing a set of things, in this case `string`s. The magic is that the map key type is the empty struct, which has zero size. Therefore the map contains the keys but no values.

chezmoi actually already has a generic implementation of sets that you can use in [`pkg/chezmoi/set.go`](https://github.com/twpayne/chezmoi/blob/36eec23ecd852a34d035a3f1ac5fde19c7550daf/pkg/chezmoi/set.go). With this you can write:

```go
    seenDirs := newSet[string]()
```

and then use the `seenDirs.add` and `seenDirs.contains` methods.